README.md
[![Supported Foundry Versions](https://img.shields.io/endpoint?url=https://foundryshields.com/version?url=https://github.com/LeafWulf/flying-tokens/releases/latest/download/module.json)](https://foundryvtt.com/packages/flying-tokens) [![Forge Installs](https://img.shields.io/badge/dynamic/json?label=Forge%20Installs&query=package.installs&suffix=%25&url=https%3A%2F%2Fforge-vtt.com%2Fapi%2Fbazaar%2Fpackage%2Fflying-tokens&colorB=0374b5)](https://forge-vtt.com/bazaar#sort=updated&package=flying-tokens)  
[![Latest Downloads](https://img.shields.io/github/downloads/paulo-roger/flying-tokens/latest/module.zip?color=blue&label=latest%20downloads)](https://github.com/LeafWulf/flying-tokens/releases/latest) [![Total Downloads](https://img.shields.io/github/downloads/paulo-roger/flying-tokens/module.zip?color=blue&label=total%20downloads)](https://github.com/LeafWulf/flying-tokens/releases)  
[![Discord](https://dcbadge.vercel.app/api/shield/219289132235489280?style=flat)](https://discordapp.com/users/219289132235489280) [![Ko-fi](https://img.shields.io/badge/Ko--fi-winterwulf-0374b5?logo=kofi)](https://ko-fi.com/winterwulf)

# Flying Tokens

Ever wanted your tokens to fly? I mean really look like they're flying, when they are supposed to, well you should try Flying Tokens.
Flying Tokens works with [Token Magic FX](https://foundryvtt.com/packages/tokenmagic) API to add shadow and movement that simulates the flight.

## Functions

With the **Automatic Apply** enabled you just have to change the token elevation and it will do all the magic. The module first check if the character has flight movement, otherwise it won't apply the effect.  
When you change the elevation to 0 the token will land.

If you opt to disable the automatic effect and prefer to enable per token you just have to disable **Automatic Apply** in the settings and now all tokens with flying speed movement will have a dove icon in the token HUD, if you click on it, it activates ***Flying Tokens*** for that specific token and then when you change its elevation the effects will apply.

***First click the dove icon to activate *Flying Tokens* for that token and then set a new elevation***  
![functions1](readme/functions1.png)

***If you want to land your token just click the dove slashed icon and it will remove the filter effects, this wont change token elevation, as for example the token could land in a higher level. Then you can change the elevation and it won't apply effects anymore.***  
![functions2](readme/functions2.png)

### Effects included
Shadow projection, token scale up or down depending on the altitude, small movements to simulate flight, Token Magic FX filter to simulate wind. In the next release it will be configurable.

## Settings
**Automatic Apply**  
*Default: active.*  
This setting if checked will apply ***Flying Tokens*** effects to all tokens that have *flying speed movement* and you change its elevation.

**Al Tokens Can Fly**  
*Default: disabled.*  
Apply ***Flying Tokens*** effects to all tokens independent of *flying speed movement* and you change its elevation.

**Output to chat**  
*Default: active.*  
Display a chat message with the token name and its new elevation.

**Notifications**  
*Default: disabled.*  
Display a notification with the token name and its new elevation.

**Scale Token Based On Elevation**  
*Default: active.*  
increase token scale based on how high is its elevation.

**Auto Zoom**  
*Default: disabled.*  
Zoom into the token when it lands, and out based on how high it is flying.

![settings](readme/settings.png)

## Dependencies
- [Token Magic FX](https://foundryvtt.com/packages/tokenmagic)

## Known Issues
So far none.

## Feedback
If you have suggestions or want to report a problem, you can create an issue here: [Issues](../../issues)

## Changelog
You can read the changelog here: [CHANGELOG.md](https://github.com/paulo-roger/flying-tokens/blob/master/CHANGELOG.md)

## Special Thanks
`Peterson`, `roi007leaf` and `Zhell`

## Donations
The module is totally free and will remain this way.  
I am unemployed, though. So every little help counts.

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/winterwulf)

## Attributions
--


CHANGELOG.md
# Changelog

## [1.3.0]
- New token scale formula, with adjustable pace.
- Now the user can set a alternative image to replace when the token is flying.
- Totally compatible with PF2E 4.2

## [1.2.1]
- Added options to select which filter effects are applied when flying. 
Hint: These settings will only affect tokens flying after the change, to apply to a token already flying you must land and fly again. 
- Fix: No longer replace the token scale but rather increase it.

## [1.1.1]
**Bugs**
- Fixed bug when setting negative elevation values (Thank you Roi!)
- Hud button re-activates flying for the current elevation
  
**Features**
- More settings: 
  - You can disable game notifications,
  -  disable auto zoom, 
  -  enable *Flying Tokens* for all tokens 
  -  and disable scale

## [1.0.0]
- Intial release

scripts/filters.js
import { MODULE, MODULE_DIR } from "./const.js"; //import the const variables
import { elevation } from "./flying-tokens.js"

export const movement = {
    filterType: "images",
    filterId: MODULE,
    time: 100,
    nbImage: 1,
    alphaImg: 1,
    alphaChr: 0.0,
    blend: 4,
    ampX: 0.005,
    ampY: 0.005,
    animated: {
        time:
        {
            active: true,
            speed: 0.0010,
            animType: "move"
        }
    }
}
export const noShadow = {
    filterType: "zapshadow",
    filterId: MODULE,
    alphaTolerance: 0.5,
    rank: 2
}
export const wind = {
    filterType: "transform",
    filterId: MODULE,
    twRadiusPercent: 10,
    padding: 150,
    animated:
    {
        twRotation:
        {
            animType: "sinOscillation",
            val1: -(0.05 * elevation),
            val2: +(0.05 * elevation),
            loopDuration: 5000,
        }
    }
}
export const shadow = {
    filterType: "shadow",
    filterId: MODULE,
    rotation: 35,
    blur: 2,
    quality: 5,
    distance: elevation / 0.7,
    alpha: Math.min(1 / ((elevation) / 50), 0.8),
    padding: elevation * 2,
    shadowOnly: false,
    color: 0x000000,
    zOrder: 6000,
    animated:
    {
        blur:
        {
            active: true,
            loopDuration: 5000,
            animType: "syncCosOscillation",
            val1: 2,
            val2: 4
        },
        rotation:
        {
            active: true,
            loopDuration: 5000,
            animType: "syncSinOscillation",
            val1: 33,
            val2: 33 + (elevation * 0.1)
        }
    }
}
export const bounce = {
    filterType: "transform",
    filterId: MODULE,
    padding: 50,
    animated:
    {
        translationX:
        {
            animType: "sinOscillation",
            val1: -0.005,
            val2: +0.005,
            loopDuration: 9600,
        },
        translationY:
        {
            animType: "cosOscillation",
            val1: -0.005,
            val2: +0.005,
            loopDuration: 1400,
        }
    }
}

scripts/settings.js
import { MODULE } from "./const.js"

export let enableFT = true;
export let enableForAll = false;
export let chatOutput = true;
export let notificationOutput = false;
export let enableZoom = false;
export let optMovement = true;
export let optNoShadow = true;
export let optWind = true;
export let customScale = null;

export function registerSettings() {
    game.settings.register(MODULE, 'enableFT', {
        name: 'Automatic Apply',
        hint: `Check this option if you would like to have tokens with flight movement automatically fly when increasing their elevation. If this setting is disabled you can still have Flying Tokens to work on individual tokens.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: true,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'enableForAll', {
        name: 'All Tokens Can Fly',
        hint: `Any token can fly, enable this if you want to make any token be able to fly.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: false,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'chatOutput', {
        name: 'Output To Chat',
        hint: `Display Flying Tokens messages in the game chat.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: true,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'notificationOutput', {
        name: 'Notifications',
        hint: `Display Flying Tokens notifications.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: false,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'customScale', {
        name: 'Custom Scale',
        hint: `Customize the escalation increment variable. The formula is: scale = originalScale + customScale * elevation. If you want to disable it set the value to 0.`,
        scope: 'world',
        config: true,
        type: Number,
        range: {
            min: 0,
            max: 0.02,
            step: 0.001
          },
        default: 0.01,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'enableZoom', {
        name: 'Auto Zoom',
        hint: `Change the zoom level based on the token scale, will only have effect if 'Scale Token Base On Elevation' is enabled.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: false,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'optMovement', {
        name: 'Enable Movement Effect.',
        hint: `This setting will only affect tokens flying after the change, to apply to a token already flying you must land and fly again.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: true,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'optNoShadow', {
        name: 'Enable Shadow Removal',
        hint: `Flying Tokens auto remove the shadow drawing from the token, uncheck this option if you want to disable this behavior. This setting will only affect tokens flying after the change, to apply to a token already flying you must land and fly again.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: true,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });

    game.settings.register(MODULE, 'optWind', {
        name: 'Enable wind Effect',
        hint: `Uncheck this to remove the small twist effect from the center of the token while it is flying. This setting will only affect tokens flying after the change, to apply to a token already flying you must land and fly again.`,
        scope: 'world',
        config: true,
        type: Boolean,
        default: true,
        restricted: true,
        onChange: () => {
            cacheSettings();
        },
    });
}

// function that get the settings options and assign to the variables
export function cacheSettings() {
    enableFT = game.settings.get(MODULE, 'enableFT');
    enableForAll = game.settings.get(MODULE, 'enableForAll');
    chatOutput = game.settings.get(MODULE, 'chatOutput');
    notificationOutput = game.settings.get(MODULE, 'notificationOutput');
    customScale = game.settings.get(MODULE, 'customScale');
    enableZoom = game.settings.get(MODULE, 'enableZoom');
    optMovement = game.settings.get(MODULE, 'optMovement');
    optNoShadow = game.settings.get(MODULE, 'optNoShadow');
    optWind = game.settings.get(MODULE, 'optWind');
}


scripts/flying-hud.js
import { MODULE, MODULE_DIR } from "./const.js"; //import the const variables
// import { getCanvas } from "./util.js";
import { isFlyer, land, fly } from "./flying-tokens.js";

/**
 * Functinality class for managing the token HUD
 */
export class FlyingHud {

    /**
     * Called when a token is right clicked on to display the HUD.
     * Adds a button with an icon, and adds a slash on top if it is disabled.
     * @param {Object} app - the application data
     * @param {Object} html - the html data
     * @param {Object} hudData - The HUD Data
     */
    static async renderHud(app, html, hudData) {
        let token = canvas.tokens.get(hudData._id)
        let enableFlight = token.document.getFlag(MODULE, "enableFlight")
        if (isFlyer(token)) {
            if (enableFlight) {
                this.addDisableButton(html, hudData)
            } else {
                this.addEnableButton(html, hudData);
            }
        }
    }
    static async addEnableButton(html, hudData) {
        // let token = canvas.tokens.controlled.filter(token => token.id == hudData._id).map(token => { return token; });
        let token = canvas.tokens.get(hudData._id)
        const button = this.buildButton(html, `Enable flight`);
        button.find('i').on("click", async (ev) => {
            await token.document.setFlag(MODULE, "enableFlight", true)
            await fly(token.document, token.document.elevation)
            canvas.hud.token.render();
        });
    }
    static async addDisableButton(html, hudData) {
        let token = canvas.tokens.get(hudData._id)
        let button = this.buildButton(html, `Disable flight`);
        button = this.addSlash(button);

        button.find('i').on("click", async (ev) => {
            land(token.document)
            await token.document.setFlag(MODULE, "enableFlight", false)
            canvas.hud.token.render();
        });
    }

    static buildButton(html, tooltip) {
        let button = $(`<div class="control-icon mount-up" title="${tooltip}"><i class="fas fa-dove"></i></div>`);
        let col = html.find('.col.left');
        col.prepend(button);
        return button;
    }
    /**
     * Adds a slash icon on top of the horse icon to signify "dismount"
     * @param {Object} button - The HUD button to add a slash on top of
     */
    static addSlash(button) {
        let slash = $(`<i class="fas fa-slash" style="position: absolute; color: grey"></i>`);
        button.addClass("fa-stack");
        button.find('i').addClass('fa-stack-1x');
        slash.addClass('fa-stack-1x');
        button.append(slash);
        return button;
    }
}



scripts/flying-tokens.js
import { MODULE, MODULE_DIR } from "./const.js"; //import the const variables
import { chatMessage } from "./util.js"
import { registerSettings, cacheSettings, enableFT, enableForAll, enableZoom, chatOutput, notificationOutput, optMovement, optNoShadow, optWind, customScale } from "./settings.js" //import settings
import { FlyingHud } from "./flying-hud.js"
// import { movement, noShadow, wind, shadow, bounce } from "./filters.js"

//Compatibility with v9
let fvttVersion
//Compatibility with PF2E
let system

// Hook that trigger once when the game is initiated. Register and cache settings.
Hooks.once("init", () => {
    // registerWrappers();
    registerSettings();
    cacheSettings();
});

Hooks.once('ready', async function () {
    fvttVersion = parseInt(game.version)
    system = game.system.id
    console.log(" ====================================== 🐦 Flying Tokens  ======================================= ")
    console.log(" ==================================== FoundryVTT Version:", fvttVersion, " ==================================== ")
    pf2eAutoScaleCheck();
});

function pf2eAutoScaleCheck(token, permanent = true) {
    let check;
    if (system == 'pf2e' && customScale) {
        if (token) check = token.getFlag("pf2e", "linkToActorSize")
        else check = game.settings.get('pf2e', 'tokens.autoscale')
        if (check) ui.notifications.warn('If you want Flying Tokens to autoscale you must disable PF2E setting "<b>Scale tokens according to size</b>" or individually disable this in the token config.', { permanent })
    }
}

Hooks.on("preUpdateToken", async (token, updateData) => {
    // await token.setFlag(MODULE, "scale", token.texture.scaleX);
    let enableFlight = token.getFlag(MODULE, "enableFlight")
    if (enableFT || enableFlight) {
        let elevation = getProperty(updateData, "elevation");
        if (elevation !== undefined && isFlyer(token)) {
            await fly(token, elevation)
        }
    }
});

Hooks.on('renderTokenHUD', (app, html, data) => {
    if (!enableFT)
        FlyingHud.renderHud(app, html, data);
});

Hooks.on("renderTokenConfig", (app, html, data) => {
    if (isFlyer(app.token)) {
        if (fvttVersion >= 10) {
            let altToken = app.token.getFlag(MODULE, "altToken") || "";
            let newHtml = `<div class="form-group">
                    <label>Flying Image Path</label>
                    <div class="form-fields">
                      <button type="button" class="file-picker" data-type="imagevideo" data-target="flags.${MODULE}.altToken" title="Browse Files" tabindex="-1">
                        <i class="fas fa-file-import fa-fw"></i>
                      </button>
                      <input class="image" type="text" name="flags.${MODULE}.altToken" placeholder="path/image.png" value="${altToken}">
                    </div>
                  </div>`
            const tinthtml = html.find('input[name="texture.src"]');
            const formGroup = tinthtml.closest(".form-group");
            formGroup.after(newHtml);
            // altToken = html.find('input[name="flags.flyingTokens.alt"]').value;
            app.setPosition({ height: "auto" });
            html.find(`button[data-target="flags.${MODULE}.altToken"]`).on('click', async () => {
                new FilePicker({
                    current: altToken,
                    type: "imagevideo",
                    displayMode: "tiles",
                    button: "file-picker",
                    callback: async (path) => {
                        html.find(`input[name="flags.${MODULE}.altToken"]`).val(path);
                        // console.log(app.token)
                        // await app.token.setFlag(MODULE, "altToken", path)
                    }
                }).render()
            });
        } else {
            let altToken = app.token.getFlag(MODULE, "altToken") || "";
            let newHtml = `<div class="form-group">
                    <label>Flying Image Path</label>
                    <div class="form-fields">
                      <button type="button" class="file-picker" data-type="imagevideo" data-target="flags.${MODULE}.altToken" title="Browse Files" tabindex="-1">
                        <i class="fas fa-file-import fa-fw"></i>
                      </button>
                      <input class="image" type="text" name="flags.${MODULE}.altToken" placeholder="path/image.png" value="${altToken}">
                    </div>
                  </div>`
            const tinthtml = html.find('input[name="img"]');
            const formGroup = tinthtml.closest(".form-group");
            formGroup.after(newHtml);
            // altToken = html.find('input[name="flags.flyingTokens.alt"]').value;
            app.setPosition({ height: "auto" });
            html.find(`button[data-target="flags.${MODULE}.altToken"]`).on('click', async () => {
                new FilePicker({
                    current: altToken,
                    type: "imagevideo",
                    displayMode: "tiles",
                    // button: "file-picker",
                    callback: async (path) => {
                        html.find(`input[name="flags.${MODULE}.altToken"]`).val(path);
                        // console.log(app.token)
                        // await app.token.setFlag(MODULE, "altToken", path)
                    }
                }).render()
            });
        }
    }
});

export function isFlyer(token) {
    if (enableForAll) return true;
    let tokenFly = token.actor.data.data.attributes.movement.fly
    if (tokenFly <= 0 || tokenFly == null) {
        let errorMessage = "This creature can't fly."
        // ui.notifications.error(errorMessage);
        console.log("Flying Token error: ", errorMessage)
        return false;
    }
    else return true;
}

export async function fly(token, elevation) {
    let scale
    if (fvttVersion >= 10)
        scale = token.texture.scaleX;
    if (fvttVersion < 10)
        scale = token.data.scale
    let isFlying = token.getFlag(MODULE, "flying")
    if (!isFlying) {
        await token.setFlag(MODULE, "scale", scale);
        if (fvttVersion < 10) await token.setFlag(MODULE, "originalToken", token.data.img);// compatible v9
        else await token.setFlag(MODULE, "originalToken", token.texture.src);//compatible v10+
    }
    if (elevation == 0) {
        return land(token)
    } else if (elevation < 0) {
        return;
    } else {
        await token.setFlag(MODULE, "flying", true)
        await flyZoom(token, elevation)
        await flyingFX(token, elevation)
        if (notificationOutput)
            ui.notifications.info(token.data.name + ' is flying at <b>' + elevation + ' feet</b> high.')
        if (chatOutput) {
            if (fvttVersion < 10) await chatMessage(`<img src='${token.data.img}' width='32' style='border:none'> ${token.data.name} is flying at <b>${elevation} feet</b> high.`)// compatible v9
            else await chatMessage(`<img src='${token.texture.src}' width='32' style='border:none'> ${token.name} is flying at <b>${elevation} feet</b> high.`)//compatible v10+
        }
    }
}

async function tokenScale(token, elevation) {
    pf2eAutoScaleCheck(token, false);
    let originalScale = token.getFlag(MODULE, "scale");
    let altToken = token.getFlag(MODULE, "altToken") || token.getFlag(MODULE, "originalToken");
    if (elevation == 0) return 0
    let scale = originalScale + customScale * elevation
    scale = Math.round((scale + Number.EPSILON) * 100) / 100 //rounding with 2 floats
    if (scale < 0.2) scale = 0.2
    else if (scale > 10) scale = 10
    if (fvttVersion < 10) await token.update({ img: altToken })// compatible v9
    else await token.update({ texture: { src: altToken } })//compatible v10+
    if (fvttVersion >= 10)  await new Promise(resolve => setTimeout(resolve, 800));//give time to the scale animation to play
    if (fvttVersion < 10) await token.update({ scale: scale })// compatible v9
    else await token.update({ texture: { scaleX: scale, scaleY: scale } })//this is in a different line so the animation plays AFTER the token is changed. compatible v10+
    return scale
}

async function flyZoom(token, elevation, minZoom = 3) {
    if (customScale) {
        let scale = await tokenScale(token, elevation)
        if (enableZoom) {
            let x = token.x + game.scenes.viewed.data.size
            let y = token.y + game.scenes.viewed.data.size
            let zoom = Math.min(3 / (Math.max(scale / 1.5, 1)), minZoom)
            await canvas.animatePan({ x: x, y: y, scale: zoom })
        }
    }
}

async function flyingFX(token, elevation) {
    let canvasToken = canvas.tokens.get(token.id)
    const movement = {
        filterType: "images",
        filterId: MODULE,
        time: 100,
        nbImage: 1,
        alphaImg: 1,
        alphaChr: 0.0,
        blend: 4,
        ampX: 0.005,
        ampY: 0.005,
        animated: {
            time:
            {
                active: true,
                speed: 0.0010,
                animType: "move"
            }
        }
    }
    const noShadow = {
        filterType: "zapshadow",
        filterId: MODULE,
        alphaTolerance: 0.5,
        rank: 2
    }
    const wind = {
        filterType: "transform",
        filterId: MODULE,
        twRadiusPercent: 10,
        padding: 150,
        animated:
        {
            twRotation:
            {
                animType: "sinOscillation",
                val1: -(2 * Math.pow(elevation, 0.28)),
                val2: +(2 * Math.pow(elevation, 0.28)),
                loopDuration: 5000,
            }
        }
    }
    const shadow = {
        filterType: "shadow",
        filterId: MODULE,
        rotation: 35,
        blur: 2,
        quality: 5,
        distance: elevation / 0.7,
        alpha: Math.min(1 / ((elevation) / 60), 0.8),
        padding: elevation * 2,
        shadowOnly: false,
        color: 0x000000,
        zOrder: 6000,
        animated:
        {
            blur:
            {
                active: true,
                loopDuration: 5000,
                animType: "syncCosOscillation",
                val1: 2,
                val2: 4
            },
            rotation:
            {
                active: true,
                loopDuration: 5000,
                animType: "syncSinOscillation",
                val1: 33,
                val2: 33 + (3 * Math.pow(elevation, 0.28))
            }
        }
    }
    const bounce = {
        filterType: "transform",
        filterId: MODULE,
        padding: 50,
        animated:
        {
            translationX:
            {
                animType: "sinOscillation",
                val1: -0.005,
                val2: +0.005,
                loopDuration: 9600,
            },
            translationY:
            {
                animType: "cosOscillation",
                val1: -0.005,
                val2: +0.005,
                loopDuration: 1400,
            }
        }
    }
    let flyingFXParams = []

    if (optNoShadow)
        flyingFXParams.push(noShadow, shadow)
    if (optWind)
        flyingFXParams.push(wind)
    if (optMovement)
        flyingFXParams.push(bounce, movement)

    let isFlying = token.getFlag(MODULE, "flying")
    await canvasToken.TMFXdeleteFilters(MODULE)
    if (isFlying)
        await TokenMagic.addUpdateFilters(canvasToken, flyingFXParams);
}

export async function land(token) {
    await token.setFlag(MODULE, "flying", false);
    let originalToken = token.getFlag(MODULE, "originalToken");
    let scale = token.getFlag(MODULE, "scale");
    if (fvttVersion < 10) await token.update({ scale: scale })// compatible v9
    else await token.update({ texture: { scaleX: scale, scaleY: scale } })//compatible v10+
    await flyZoom(token, 0, 2.5);
    if (fvttVersion >= 10)  await new Promise(resolve => setTimeout(resolve, 800));//give time to the scale animation to play
    await flyingFX(token, 0);
    if (fvttVersion < 10) await token.update({ img: originalToken })// compatible v9
    else await token.update({ texture: { src: originalToken } }) //this is in a different line so the animation plays BEFORE the token is changed. compatible v10+
    if (notificationOutput)
        ui.notifications.info(token.data.name + ' <b> has landed</b>.')
    if (chatOutput) {
        if (fvttVersion < 10) await chatMessage(`<img src='${token.data.img}' width='32' style='border:none'> ${token.data.name} <b> has landed</b>.`)// compatible v9
        else await chatMessage(`<img src='${token.texture.src}' width='32' style='border:none'> ${token.name}  <b> has landed</b>.`)//compatible v10+
    }
}
